text/0456-elasticache-l2.md

```diff
@@ -0,0 +1,374 @@
+Amazon ElastiCache CDK module (`@aws-cdk/aws-elasticache`). This launch supports
+Amazon ElastiCache Redis replication groups (distributed in-memory data store
+or cache environment) within the CDK.
+
+### Why should I use this feature?
+
+Setting up AWS ElastiCache with AWS CDK minimizes work needed to provide access
+to other services and manage distributed in-memory data store or cache
+environments in the cloud.
+
+## Internal FAQ
+
+### Why are we doing this?
+
+There is a community offer to support with implementation right now.
+
+The [tracking Github issue for the module](https://github.com/aws/aws-cdk/issues/6908) is
+in the top 10 upvoted issues (+123) with [previous approaches to implement an L2](https://github.com/aws/aws-cdk/issues/8196)
+that were abandoned.
+
+Setting up a CacheCluster requires setting up separate SubnetGroups that could
+actually be fully automated away in CDK. Managing VPCs and SecurityGroups
+becomes significant easier compared to manual setup or setup with CloudFormation.
+
+### Why should we *not* do this?
+
+The API design is based on one customer implementation only, so it might be
+missing insights from other use cases.
+
+There are currently not enough resources to implement the full design, only
+part of it. Implementing the rest of the design later on might lead to
+information loss between now and then.
+
+### What is the technical solution (design) of this feature?
+
+- `IEndpoint` -- interface to manage addresses and ports
+
+  ```ts
+    interface IEndpoint {
+      address: string;
+      port: number;
+    }
+  ```
+
+- `IRedisReplicationGroup` -- interface for defined and imported ReplicationGroups
+  (cluster mode disabled)
+
+  ```ts
+  interface IRedisReplicationGroup extends
+      // Since RedisReplicationGroup will extend Resource
+      cdk.IResource,
+      // To open network conns between ElastiCache and services that should access it
+      ec2.IConnectable {
+    readonly primaryEndpoint: IEndpoint;
+    readonly readerEndpoint: IEndpoint;
+    allowReaderAccessFrom(peer: IConnectable, description?: string): void;
+    allowPrimaryAccessFrom(peer: IConnectable, description?: string): void;
+    grant(grantee: iam.IGrantable, ...actions: string[]): iam.Grant;
+    metric(metricName: string, props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+    // Some canned metrics as well like `metricCPUUtilization`
+  }
+  ```
+
+- `IRedisClusterReplicationGroup` -- interface for defined and imported ReplicationGroups
+  (cluster mode enabled)
+
+  ```ts
+  interface IRedisClusterReplicationGroup extends IRedisReplicationGroup {
+    readonly configurationEndpoint: IEndpoint;
+    allowConfigurationAccessFrom(peer: IConnectable, description?: string): void;
+  }
+  ```
+
+- `RedisReplicationGroupBase` -- abstract base ReplicationGroup class with some helper
+  props/methods
+
+  ```ts
+  enum CacheNodeType {
+    M6G_LARGE = 'cache.m6g.large',
+    // ...
+  }
```

Comment:
Should the list of of node types be defined explicitly, or can we use `ec2.InstanceType` and add a prefix like the [RDS module does](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-rds/lib/cluster.ts#L951-L953)

It would mean that users would get a deploy time error for using an instance type not supported by ElastiCache, but even with this explicit list, some CacheNodeTypes won't be available in all regions or supported by all Redis versions.

It means you could define an invalid instance type without type checking

Reply:
Makes sense! Will adapt the RFC unless I hear arguments against this.

Reply:
I've adapted this but it turned out to be slightly different from RDS, as for RDS there are multiple instance settings which don't fully translate to ElastiCache.